File: src/main/scala/de/theunknownxy/jgui/container/SingleContainer.scala
package de.theunknownxy.jgui.container

import de.theunknownxy.jgui.base.{Widget, Point}
import de.theunknownxy.jgui.event.MouseButton.MouseButton

abstract class SingleContainer extends Container {
  var child: Option[Widget] = None

  def setChild(widget: Widget): Unit = {
    child = Some(widget)
  }

  override def mouseClick(pos: Point, button: MouseButton): Option[Widget] = {
    child filter(_.area contains pos) flatMap(_.mouseClick(pos, button))
  }

  override def draw(): Unit = {
    child map (_ draw())
  }
}

File: src/main/kotlin/de/theunknownxy/jgui/container/SingleContainer.kt
package de.theunknownxy.jgui.container

import de.theunknownxy.jgui.base.Widget
import de.theunknownxy.jgui.base.Point
import de.theunknownxy.jgui.event.MouseButton

public abstract class SingleContainer : Container() {
    public var child: Widget? = null

    override fun onMouseClick(pos: Point, button: MouseButton): Widget? {
        val c = child
        if (c != null && c.rect.contains(pos)) {
            return c.onMouseClick(pos, button)
        }
        return null
    }

    override fun onUpdate() {
        child?.onUpdate()
    }

    override fun draw() = child?.draw()
}